## Импортируем необходимые библиотеки.

In [1]:
import pandas as pd
import numpy as np

## Прочитаем датасет в переменную.

In [2]:
df = pd.read_csv('cian_dataset.csv')
pd.options.display.max_columns = None

## Проанализируем датасет.

In [3]:
df.head()

Unnamed: 0  ID  объявления Количество комнат       Тип  \
0           0       271271157                 4  Квартира   
1           1       271634126                 4  Квартира   
2           2       271173086   4, Оба варианта  Квартира   
3           3       272197456   4, Оба варианта  Квартира   
4           4       273614615                 2  Квартира   

                          Метро                             Адрес  \
0  м. Смоленская (9 мин пешком)     Москва, улица Новый Арбат, 27   
1  м. Смоленская (8 мин пешком)     Москва, улица Новый Арбат, 27   
2  м. Смоленская (7 мин пешком)     Москва, улица Новый Арбат, 27   
3  м. Смоленская (3 мин пешком)  Москва, переулок Плотников, 21С1   
4   м. Арбатская (7 мин пешком)     Москва, улица Новый Арбат, 15   

       Площадь, м2                        Дом   Парковка  \
0       200.0/20.0           5/16, Монолитный  подземная   
1  198.0/95.0/18.0  5/16, Монолитно-кирпичный  подземная   
2  200.0/116.0/4.0                       5/16  подземная   
3  170.0/95.0/17.0                        5/6  подземная   
4    58.0/38.0/5.0           12/26, Панельный        NaN   

                                                Цена  \
0  500000.0 руб./ За месяц, Залог - 500000 руб., ...   
1  500000.0 руб./ За месяц, Залог - 500000 руб., ...   
2  500000.0 руб./ За месяц, Залог - 500000 руб., ...   
3  400000.0 руб./ За месяц, Залог - 400000 руб., ...   
4  225000.0 руб./ За месяц, Залог - 225000 руб., ...   

                     Телефоны  \
0                +79166369231   
1                +79850608590   
2  +79672086536, +79099269384   
3  +79660342340, +79099269384   
4                +79852432860   

                                            Описание        Ремонт  \
0  Без комиссии для нанимателя! Бонус коллегам 12...  Дизайнерский   
1  Лот 93107. Елена Анисимова.\n\nБонус агенту 50...  Дизайнерский   
2  ID 36380:  Шикарная 4-х км. квартира в ЖК "Нов...    Евроремонт   
3  ID 31618:  Эксклюзивное предложение. Современн...    Евроремонт   
4  Лот 111542. Татьяна Лучкина.\n\nБонус агенту 5...    Евроремонт   

  Площадь комнат, м2 Балкон             Окна                          Санузел  \
0                NaN    NaN              NaN                              NaN   
1        25 25 20 25    NaN  На улицу и двор  Совмещенный (2), Раздельный (1)   
2                NaN    NaN  На улицу и двор                  Совмещенный (3)   
3        14-42-20-19    NaN  На улицу и двор                  Совмещенный (3)   
4              20 18    NaN  На улицу и двор                  Совмещенный (2)   

            Можно с детьми/животными  \
0  Можно с детьми, Можно с животными   
1                     Можно с детьми   
2                     Можно с детьми   
3                  Можно с животными   
4                                NaN   

                                       Дополнительно        Название ЖК  \
0  Мебель в комнатах, Мебель на кухне, Ванна, Душ...  Новый Арбат, 2010   
1  Мебель в комнатах, Мебель на кухне, Ванна, Душ...        Новый Арбат   
2  Мебель в комнатах, Мебель на кухне, Ванна, Душ...        Новый Арбат   
3  Мебель в комнатах, Мебель на кухне, Ванна, Душ...                NaN   
4  Мебель в комнатах, Мебель на кухне, Ванна, Душ...           The Book   

  Серия дома  Высота потолков, м                Лифт Мусоропровод  \
0        NaN                 3.0  Пасс (4), Груз (1)           Да   
1        NaN                 3.5  Пасс (1), Груз (1)          Нет   
2        NaN                 3.2            Пасс (1)          NaN   
3        NaN                 3.2            Пасс (1)          NaN   
4        NaN                 3.9  Пасс (1), Груз (1)           Да   

                      Ссылка на объявление  
0  https://www.cian.ru/rent/flat/271271157  
1  https://www.cian.ru/rent/flat/271634126  
2  https://www.cian.ru/rent/flat/271173086  
3  https://www.cian.ru/rent/flat/272197456  
4  https://www.cian.ru/rent/flat/273614615

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23368 entries, 0 to 23367
Data columns (total 25 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                23368 non-null  int64  
 1   ID  объявления            23368 non-null  int64  
 2   Количество комнат         22327 non-null  object 
 3   Тип                       23368 non-null  object 
 4   Метро                     22053 non-null  object 
 5   Адрес                     23368 non-null  object 
 6   Площадь, м2               23368 non-null  object 
 7   Дом                       23368 non-null  object 
 8   Парковка                  9951 non-null   object 
 9   Цена                      23368 non-null  object 
 10  Телефоны                  23368 non-null  object 
 11  Описание                  23368 non-null  object 
 12  Ремонт                    20613 non-null  object 
 13  Площадь комнат, м2        14458 non-null  object 
 14  Балкон

In [5]:
df.isna().sum()

Unnamed: 0                      0
ID  объявления                  0
Количество комнат            1041
Тип                             0
Метро                        1315
Адрес                           0
Площадь, м2                     0
Дом                             0
Парковка                    13417
Цена                            0
Телефоны                        0
Описание                        0
Ремонт                       2755
Площадь комнат, м2           8910
Балкон                       7978
Окна                         6613
Санузел                      2672
Можно с детьми/животными     6096
Дополнительно                 357
Название ЖК                 17520
Серия дома                  21205
Высота потолков, м          12162
Лифт                         5500
Мусоропровод                10522
Ссылка на объявление            0
dtype: int64

In [6]:
df.describe()

Unnamed: 0  ID  объявления  Высота потолков, м
count  23368.000000    2.336800e+04         11206.00000
mean   11683.500000    2.676265e+08             2.98345
std     6745.904881    1.876862e+07             7.61472
min        0.000000    1.072986e+08             1.20000
25%     5841.750000    2.712663e+08             2.64000
50%    11683.500000    2.739030e+08             2.64000
75%    17525.250000    2.746872e+08             2.80000
max    23367.000000    2.750064e+08           320.00000

Нужно:
- Unnamed: 0 - удалить
- ID  объявления - удалить
- Количество комнат - избавиться от слов и перевести в тип int (например "4, Оба варианта"), 1041 пропущенных значений
- Тип - удалить, так как одно единственное значение - квартира;
- Метро - разбить на 2 столбца, разделить метро и расстояние до него, 1315 пропущенных значений;
- Адрес - пока выглядит ок
- Площадь, м2 - разбить на три столбца
- Дом - разбить на три столбца (текущий этаж, общая этажность, тип дома);
- Парковка - 13417 пропущенных значений;
- Цена - подумать как разбить
- Телефоны - удалить
- Описание - пока выглядит ок
- Ремонт - 2755 пропущенных значений;
- Площадь комнат, м2 - привести в единый формат, 8910 пропущенных значений;
- Балкон - разбить на балкон и лоджию (пример: "Балкон (1), Лоджия (1)"), 7978 пропущенных значений;
- Окна - разбить на улицу и двор, 6613 пропущенных значений;
- Санузел - разбить на совмещенный и раздельный, 2672 пропущенных значений;
- Можно с детьми/животными - разбить на 3, 6096 пропущенных значений;
- Дополнительно - слишком много всего, что можно вытащить, но удалять жалко, 357 пропущенных значений;
- Название ЖК - удалить, 17520 пропущенных значений;
- Серия дома - думаю можно удалить, 21205 пропущенных значений;
- Высота потолков, м - поработать с выбросами, 12162 пропущенных значений;
- Лифт - разбить на пассажирский и грузовой, 5500 пропущенных значений;
- Мусоропровод - 10522 пропущенных значений;
- Ссылка на объявление - удалить.       

## Работа над датасетом.

### 1. Удалим ненужные столбцы:

In [7]:
df.drop(columns=['Unnamed: 0', 'ID  объявления', 'Тип', 'Телефоны', 'Название ЖК', 'Серия дома', 'Ссылка на объявление'], axis=1, inplace=True)

### 2. Количество комнат: 

#### 2.1. Удалим лишние слова, и оставим только количество

In [8]:
df['Количество комнат'].value_counts()

1                   7917
2, Изолированная    4623
2                   2591
3                   1717
3, Изолированная    1583
3, Оба варианта      875
4                    674
2, Смежная           637
2, Оба варианта      615
4, Оба варианта      253
5                    235
4, Изолированная     223
6                     87
3, Смежная            87
5, Оба варианта       81
5, Изолированная      47
6, Оба варианта       31
6, Изолированная      17
4, Смежная            13
1, Изолированная       8
1, Оба варианта        4
5, Смежная             4
6, Смежная             3
1, Смежная             2
Name: Количество комнат, dtype: int64

In [9]:
df['Количество комнат'] = df['Количество комнат'].astype('str')
df['Количество комнат'].dtype

dtype('O')

In [10]:
df[['Количество комнат', 'Тип комнат']] = df['Количество комнат'].str.split(',', expand=True)
df.drop(columns='Тип комнат', inplace=True)

#### 2.2. Обработаем пропуски:

In [11]:
df['Количество комнат'].value_counts().sort_values()

6       138
5       367
nan    1041
4      1163
3      4262
1      7931
2      8466
Name: Количество комнат, dtype: int64

Посмотрим, какой площади соответствуют данные квартиры:

In [12]:
df[df['Количество комнат'] == 'nan']['Площадь, м2']

115                41.0
120                63.0
270            26.0/5.0
288                27.0
289       27.7/13.4/6.8
              ...      
23334              23.0
23348              35.0
23353    45.0/22.0/15.0
23362     25.0/17.0/4.0
23363     35.0/16.4/8.0
Name: Площадь, м2, Length: 1041, dtype: object

Предположим, что nan имеет отношение к студиям и заменим пропущенные значения на 0.

In [13]:
df['Количество комнат'] = df['Количество комнат'].replace('nan', '0')
#df['Количество комнат'].value_counts().sort_values()

### 3. Метро:

Разделим на два столбца 'Метро' и 'Расстояние до метро, мин пешком':

In [14]:
df[['Метро', 'Расстояние до метро, мин пешком']] = df['Метро'].str.split('(', expand=True)
df['Метро'] = df['Метро'].str.strip('м. ')
df['Расстояние до метро, мин пешком'] = df['Расстояние до метро, мин пешком'].str[0]

In [15]:
df['Расстояние до метро, мин пешком'].value_counts()

1    8960
5    2140
2    1884
7    1514
3    1409
4    1365
8    1283
6    1244
9    1115
N    1040
0      99
Name: Расстояние до метро, мин пешком, dtype: int64

В столбце 'Расстояние до метро, мин пешком' заменим пропущенные значения путем заполнения медианными значениями соответствущего метро.

In [16]:
df['Расстояние до метро, мин пешком'] = df['Расстояние до метро, мин пешком'].replace('N', np.nan)

In [17]:
for name in df['Метро'].unique():
    median = df.loc[(df['Метро'] == name), 'Расстояние до метро, мин пешком'].median()
    df.loc[(df['Расстояние до метро, мин пешком'].isna()) & (df['Метро'] == name), 'Расстояние до метро, мин пешком'] = median

/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/numpy/lib/nanfunctions.py:1215: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/numpy/lib/nanfunctions.py:1215: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)


Приведем столбец 'Расстояние до метро, мин пешком' к типу int:

In [19]:
df['Расстояние до метро, мин пешком'] = df['Расстояние до метро, мин пешком'].astype(int)

ValueError: cannot convert float NaN to integer